### PR TITLE
LibCompress: Make the DEFLATE implementation a true stream implemenation.

### PR DESCRIPTION
**Pull request by asynts**
In #3220 I've already tried to make this a stream implementation, but I think I took the wrong approach. It is possible to read a DEFLATE stream while only looking at 64KiB of data at a time, which this implementation now does. (Previously, I read an entire block at once which could be significantly more than 64KiB and required allocating buffers on the heap.)

There are still a few FIXMEs in this implementation but I want to do some other unrelated stuff first.

### CODE REVIEW

File: AK/Stream.h [hunk 1]
```diff
@@ -104,6 +105,22 @@ InputStream& operator<<(InputStream& stream, BigEndian<T> value)
...
+    value = temporary;
+    return stream;
+}
+template<typename T>
+OutputStream& operator<<(OutputStream& stream, const Optional<T>& value)
+{
+    if (value.has_value())
+        return stream << value.value();
+    return stream;
+}
```
Comment (awesomekling) [inline AK/Stream.h:122]: This feels like a very weird API. It may or may not write to the stream, depending of whether the optional is empty or not.

Comment (asynts) [reply AK/Stream.h]: Initially, I had the same thought but then I dismissed it, but I have to agree it's pretty weird and I can't think of a use case. I've rebased and removed it.

File: Libraries/LibCompress/Deflate.h [hunk 2]
```diff
@@ -26,160 +26,90 @@
...
+    State m_state { State::Idle };
+    union {
+        CompressedBlock m_compressed_block;
+        UncompressedBlock m_uncompressed_block;
+    };
 
-    mutable bool m_read_last_block { false };
-    mutable DuplexMemoryStream m_intermediate_stream;
+    InputBitStream m_input_stream;
+    CircularDuplexStream<64 * 1024> m_output_stream;
```
Comment (stelar7) [inline Libraries/LibCompress/Deflate.h:112]: I might be misunderstanding this piece of code, but the circular buffer used in DEFLATE should only be 32K

Comment (asynts) [reply Libraries/LibCompress/Deflate.h]: You are correct this should be 32KiB, no idea why I got 64KiB there.

File: Libraries/LibCompress/Deflate.h [hunk 3]
```diff
@@ -26,160 +26,90 @@
...
+    State m_state { State::Idle };
+    union {
+        CompressedBlock m_compressed_block;
+        UncompressedBlock m_uncompressed_block;
+    };
 
-    mutable bool m_read_last_block { false };
-    mutable DuplexMemoryStream m_intermediate_stream;
+    InputBitStream m_input_stream;
+    CircularDuplexStream<32 * 1024> m_output_stream;
```
Comment (awesomekling) [inline Libraries/LibCompress/Deflate.h:112]: I guess the commit message should be updated too? :)

Comment (asynts) [reply Libraries/LibCompress/Deflate.h:112]: done.

**Pull request merged**